cmd/tempo-cli/cmd-rewrite-blocks.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	fmt.Println("successfully rewrote blocks dropping requested trace")
+
+	return nil
+}
+
+func rewriteBlock(ctx context.Context, r backend.Reader, w backend.Writer, meta *backend.BlockMeta, traceID common.ID) (*backend.BlockMeta, error) {
+	enc, err := encoding.FromVersion(meta.Version)
+	if err != nil {
+		return nil, fmt.Errorf("error getting encoder: %w", err)
+	}
+
+	// todo: provide a way to pass a config in. this just uses defaults which is fine for now
+	opts := common.CompactionOptions{
+		BlockConfig: common.BlockConfig{
+			// defaults should be fine for just recreating a few blocks
+			BloomFP:             common.DefaultBloomFP,
+			BloomShardSizeBytes: common.DefaultBloomShardSizeBytes,
+			Version:             meta.Version,
+
+			// these fields aren't in use anymore. we need to remove the old flatbuffer search. setting them for completeness
+			SearchEncoding:      backend.EncSnappy,
+			SearchPageSizeBytes: 1024 * 1024,
+
+			// v2 fields
+			IndexDownsampleBytes: common.DefaultIndexDownSampleBytes,
+			IndexPageSizeBytes:   common.DefaultIndexPageSizeBytes,
+			Encoding:             backend.EncZstd,
+
+			// parquet fields
+			RowGroupSizeBytes: int(meta.Size / uint64(meta.TotalRecords)), // set to the average row group size of the current block
+
+			// vParquet3 fields
+			DedicatedColumns: meta.DedicatedColumns,
+		},
+		ChunkSizeBytes:     tempodb.DefaultChunkSizeBytes,
+		FlushSizeBytes:     tempodb.DefaultFlushSizeBytes,
+		IteratorBufferSize: tempodb.DefaultIteratorBufferSize,
+		OutputBlocks:       1,
+		Combiner:           model.StaticCombiner, // this should never be necessary b/c we are only compacting one block
+		MaxBytesPerTrace:   0,                    // disable for this process
+
+		// hook to drop the trace
+		DropObject: func(id common.ID) bool {
+			return bytes.Equal(id, traceID)
+		},
+
+		// setting to prevent panics. should we track and report these?
+		BytesWritten:      func(_, _ int) {},
+		ObjectsCombined:   func(_, _ int) {},
+		ObjectsWritten:    func(_, objs int) {},
+		SpansDiscarded:    func(_, _, _ string, _ int) {},
+		DisconnectedTrace: func() {},
+		RootlessTrace:     func() {},
+	}
+
+	compactor := enc.NewCompactor(opts)
+	fmt.Println("--beginning compaction logs--")
+	out, err := compactor.Compact(ctx, log.NewLogfmtLogger(os.Stdout), r, w, []*backend.BlockMeta{meta})
+	fmt.Println("--ending compaction logs--")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(out) != 1 {
+		return nil, fmt.Errorf("expected 1 block, got %d", len(out))
+	}
+
+	newMeta := out[0]
+
+	if newMeta.TotalObjects != meta.TotalObjects-1 {
+		return nil, fmt.Errorf("expected output to have one less object then in. out: %d in: %d", newMeta.TotalObjects, meta.TotalObjects)
+	}
+
+	return newMeta, nil
+}
+
+func blocksWithTraceID(ctx context.Context, r backend.Reader, tenantID string, traceID common.ID) ([]*backend.BlockMeta, error) {
+	blockIDs, compactedBlockIDs, err := r.Blocks(context.Background(), tenantID)
```

Comment:
Is it worth increasing the amount of work by picking also compacted blocks? They'll be deleted eventually anyway.

Reply:
good call. removed compactedBlockIDs